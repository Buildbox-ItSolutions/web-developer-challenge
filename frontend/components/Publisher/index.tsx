import { ChangeEvent, FC, useState } from 'react';

import { submitProps } from '../Main';
import {
    CancelButton,
    Img,
    Input,
    ProfilePicture,
    Publisher,
    SubmitButton,
    Textarea,
    Wrapper,
    WrapperButton
} from './styles';

interface Props {
    handleSubmit: (props: submitProps) => void;
}

const PublisherComponent: FC<Props> = ({ handleSubmit }) => {
    const [profilePic, setProfilePic] = useState<null | string>(null);
    const [author, setAuthor] = useState('');
    const [message, setMessage] = useState('');
    const onChange = (e: ChangeEvent<HTMLInputElement>) => {
        setProfilePic(e.target.files[0] && URL.createObjectURL(e.target.files[0]));
    };

    const handleCancel = () => {
        setProfilePic(null);
        setMessage('');
        setAuthor('');
    };

    return (
        <Publisher
            data-test="publisher-form"
            onSubmit={(e) => {
                handleCancel();
                handleSubmit({ e, message, author, profilePic });
            }}>
            <Wrapper>
                <ProfilePicture>
                    <label htmlFor="file">
                        {profilePic ? (
                            <Img src={profilePic} alt="profile" width="100%" height="100%" />
                        ) : (
                            <Img src="image.svg" alt="profile" width="24px" height="24px" />
                        )}
                    </label>
                    <input type="file" id="file" name="file" onChange={onChange} accept="image/*" />
                </ProfilePicture>
                {profilePic !== null && (
                    <Img
                        src="/trash.svg"
                        alt="trash"
                        width="24px"
                        height="24px"
                        color="#353231"
                        onClick={() => setProfilePic(null)}
                    />
                )}
            </Wrapper>
            <Input
                placeholder="Digite seu nome"
                value={author}
                onChange={(e) => setAuthor(e.target.value)}
                required
                data-test="publisher-input-name"
                name="name"
            />
            <Textarea
                placeholder="Mensagem"
                value={message}
                onChange={(e) => setMessage(e.target.value)}
                maxLength={160}
                required
                name="message"
                data-test="publisher-textarea-msg"
            />
            <WrapperButton>
                <CancelButton onClick={handleCancel} type="reset">
                    Descartar
                </CancelButton>
                <SubmitButton data-test="publisher-btn-submit" type="submit">
                    Publicar
                </SubmitButton>
            </WrapperButton>
        </Publisher>
    );
};

export default PublisherComponent;
