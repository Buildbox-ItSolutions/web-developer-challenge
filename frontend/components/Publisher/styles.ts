import styled from 'styled-components';

export const Publisher = styled.form`
    width: 80%;
    height: 353px;
    background-color: #313131;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    align-items: center;
    border-radius: 3px;
    padding: 2em;
    border: 1px solid #3b3b3b;
    grid-column-start: 2;
    grid-column-end: 2;

    @media (min-width: 900px) {
        width: 100%;
    }
`;

export const Wrapper = styled.div`
    width: 100%;
    display: flex;
    justify-content: center;
    align-items: center;
`;

export const ProfilePicture = styled.div`
    margin-right: 1em;
    width: 88px;
    height: 88px;
    border-radius: 36px;
    display: flex;
    justify-content: center;
    align-items: center;
    position: relative;
    border: 1px solid #4b4b4b;

    input {
        display: none;
    }

    label {
        z-index: 10;
        width: 100%;
        height: 100%;
        border-radius: 36px;
        display: flex;
        justify-content: center;
        align-items: center;
        cursor: pointer;
        text-align: center;
    }
`;

const defaultInputProps = `
    width: 100%;
    border-radius: 8px;
    background-color: #494949;
    border: none;
    color: #ffff;
    padding: 0.9em;
    :focus {
        outline-width: 0;
    }
`;

export const Input = styled.input`
    ${defaultInputProps}
    height: 40px;
`;

export const Textarea = styled.textarea`
    ${defaultInputProps}
    height: 80px;
    resize: none;
`;

export const WrapperButton = styled.div`
    width: 100%;
    display: flex;
    justify-content: flex-end;
`;

export const SubmitButton = styled.button`
    width: 98px;
    height: 41px;
    object-fit: contain;
    border-radius: 8px;
    background-color: #5f5f5f;
    color: #313131;
    border: none;
    transition: 0.3s;
    font-size: 14px;
    cursor: pointer;

    :focus {
        outline-width: 0;
    }

    :active {
        background-color: #71bb00;
        color: #fff;
    }
`;

export const CancelButton = styled.button`
    width: 60px;
    font-size: 14px;
    color: #5f5f5f;
    background: none;
    border: none;
    cursor: pointer;
    text-decoration: underline;
    margin-right: 1.5em;

    :focus {
        outline-width: 0;
    }
`;

export const Img = styled.img<{ width: string; height: string; color?: string }>`
    width: ${({ width }) => width};
    height: ${({ height }) => height};
    color: ${({ color }) => color};
    cursor: pointer;
    border-radius: 36px;
`;
