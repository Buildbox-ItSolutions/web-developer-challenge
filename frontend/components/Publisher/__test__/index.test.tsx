import 'jest-styled-components';

import { shallow } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';

import Publisher from '../index';

describe('Publisher', () => {
    it('must render without errors', () => {
        const handleSubmit = jest.fn();
        const wrapper = renderer.create(<Publisher handleSubmit={handleSubmit} />).toJSON();

        expect(wrapper).toMatchSnapshot();
    });

    it('calls handleSubmit prop function when form is submitted', () => {
        const handleSubmit = jest.fn();
        const wrapper = shallow(<Publisher handleSubmit={handleSubmit} />);

        wrapper.find(`[data-test='publisher-form']`).simulate('submit');

        expect(handleSubmit).toHaveBeenCalledTimes(1);
    });
});
