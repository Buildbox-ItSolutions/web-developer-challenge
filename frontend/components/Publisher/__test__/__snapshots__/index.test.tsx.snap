// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Publisher must render without errors 1`] = `
.c0 {
  width: 80%;
  height: 353px;
  background-color: #313131;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 3px;
  padding: 2em;
  border: 1px solid #3b3b3b;
  grid-column-start: 2;
  grid-column-end: 2;
}

.c1 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  margin-right: 1em;
  width: 88px;
  height: 88px;
  border-radius: 36px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  border: 1px solid #4b4b4b;
}

.c2 input {
  display: none;
}

.c2 label {
  z-index: 10;
  width: 100%;
  height: 100%;
  border-radius: 36px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  text-align: center;
}

.c4 {
  width: 100%;
  border-radius: 8px;
  background-color: #494949;
  border: none;
  color: #ffff;
  padding: 0.9em;
  height: 40px;
}

.c4:focus {
  outline-width: 0;
}

.c5 {
  width: 100%;
  border-radius: 8px;
  background-color: #494949;
  border: none;
  color: #ffff;
  padding: 0.9em;
  height: 80px;
  resize: none;
}

.c5:focus {
  outline-width: 0;
}

.c6 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c8 {
  width: 98px;
  height: 41px;
  object-fit: contain;
  border-radius: 8px;
  background-color: #5f5f5f;
  color: #313131;
  border: none;
  -webkit-transition: 0.3s;
  transition: 0.3s;
  font-size: 14px;
  cursor: pointer;
}

.c8:focus {
  outline-width: 0;
}

.c8:active {
  background-color: #71bb00;
  color: #fff;
}

.c7 {
  width: 60px;
  font-size: 14px;
  color: #5f5f5f;
  background: none;
  border: none;
  cursor: pointer;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  margin-right: 1.5em;
}

.c7:focus {
  outline-width: 0;
}

.c3 {
  width: 24px;
  height: 24px;
  cursor: pointer;
  border-radius: 36px;
}

@media (min-width:900px) {
  .c0 {
    width: 100%;
  }
}

<form
  className="c0"
  data-test="publisher-form"
  onSubmit={[Function]}
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <label
        htmlFor="file"
      >
        <img
          alt="profile"
          className="c3"
          height="24px"
          src="image.svg"
          width="24px"
        />
      </label>
      <input
        id="file"
        name="file"
        onChange={[Function]}
        type="file"
      />
    </div>
  </div>
  <input
    className="c4"
    data-test="publisher-input-name"
    name="name"
    onChange={[Function]}
    placeholder="Digite seu nome"
    required={true}
    value=""
  />
  <textarea
    className="c5"
    data-test="publisher-textarea-msg"
    maxLength={160}
    name="message"
    onChange={[Function]}
    placeholder="Mensagem"
    required={true}
    value=""
  />
  <div
    className="c6"
  >
    <button
      className="c7"
      onClick={[Function]}
      type="reset"
    >
      Descartar
    </button>
    <button
      className="c8"
      data-test="publisher-btn-submit"
      type="submit"
    >
      Publicar
    </button>
  </div>
</form>
`;
